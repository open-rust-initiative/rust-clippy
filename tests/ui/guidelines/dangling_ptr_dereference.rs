#![allow(unused)]
#![warn(clippy::dangling_ptr_dereference)]
#![feature(rustc_private)]
extern crate libc;

use core::ffi::c_void;

#[allow(non_camel_case_types)]
type size_t = usize;

extern "C" {
    fn malloc(size: size_t) -> *mut c_void;
    fn free(ptr: *mut c_void);
}

unsafe fn free_and_deref() {
    let ptr: *mut c_void = malloc(1);

    free(ptr);
    let _ = *ptr;

    let ptr_libc: *mut c_void = libc::malloc(1);
    libc::free(ptr_libc);
    println!("{:?}", *ptr_libc);

    let ptr_1: *mut c_void = malloc(1);
    free(ptr_1);
    if !ptr_1.is_null() {
        println!("{:?}", *ptr_1);
    }

    let ptr_2: *mut c_void = malloc(1);
    if !ptr_2.is_null() {
        free(ptr_2); // FN, parent block of this free is not the whole function body
    }
    if !ptr_2.is_null() {
        println!("{:?}", *ptr_2);
    }
}

// Don't lint anything inside
unsafe fn free_but_replaced() {
    let mut ptr: *mut c_void = malloc(1);

    free(ptr);
    ptr = malloc(1);
    let _ = *ptr;

    let shadow: *mut c_void = malloc(1);
    free(shadow);
    let shadow: *mut c_void = libc::malloc(1);
    let _ = *shadow;
}

fn main() {}
