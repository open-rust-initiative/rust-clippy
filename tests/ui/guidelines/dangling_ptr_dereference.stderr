error: dereferencing a raw pointer that was already freed
  --> $DIR/dangling_ptr_dereference.rs:20:13
   |
LL |     let _ = *ptr;
   |             ^^^^
   |
note: the pointer was freed here
  --> $DIR/dangling_ptr_dereference.rs:19:5
   |
LL |     free(ptr);
   |     ^^^^^^^^^
   = note: `-D clippy::dangling-ptr-dereference` implied by `-D warnings`

error: dereferencing a raw pointer that was already freed
  --> $DIR/dangling_ptr_dereference.rs:24:22
   |
LL |     println!("{:?}", *ptr_libc);
   |                      ^^^^^^^^^
   |
note: the pointer was freed here
  --> $DIR/dangling_ptr_dereference.rs:23:5
   |
LL |     libc::free(ptr_libc);
   |     ^^^^^^^^^^^^^^^^^^^^

error: dereferencing a raw pointer that was already freed
  --> $DIR/dangling_ptr_dereference.rs:29:26
   |
LL |         println!("{:?}", *ptr_1);
   |                          ^^^^^^
   |
note: the pointer was freed here
  --> $DIR/dangling_ptr_dereference.rs:27:5
   |
LL |     free(ptr_1);
   |     ^^^^^^^^^^^

error: aborting due to 3 previous errors

